File: kernel/epsilon/epsilon_core/persistent_kv_partition.py
# Epsilon-Hollow - Copyright (c) 2024 Teerth Sharma
# SPDX-License-Identifier: Epsilon-Hollow

"""
Epsilon-Hollow — Persistent Homology KV-Cache Partitioning (PHKP)
===================================================================
Author: Teerth Sharma
Status: Original Theorem — Lean 4 Verification Pending

═══════════════════════════════════════════════════════════════
THEOREM 7: Persistent Homology KV-Cache Partitioning (PHKP)
═══════════════════════════════════════════════════════════════

Uses Betti numbers to determine optimal partitioning of the KV cache
across memory tiers (HBM3 on H100, DDR5, NVMe). The theorem bounds
retrieval latency given a partition that respects the topological
structure of the key space.

Motivation:
    H100 has 80GB HBM3 (3.35 TB/s) + host DDR5 (204.8 GB/s) + NVMe
    (~7 GB/s). A 70B model's KV cache at long context (128K tokens)
    can exceed HBM capacity. WHICH keys go in HBM vs DDR vs NVMe?

    Random partitioning: some queries hit NVMe, latency spikes.
    Our approach: use the topological structure (Betti-0 clusters)
    to keep topologically connected keys together in the same tier.

Definition (Topological Locality):
    For a key space K with Betti-0 decomposition {C₁,...,C_P}, the
    topological locality of a partition π: K → {HBM, DDR, NVMe} is:

        L(π) = Σ_{p=1}^{P} |{tiers used by C_p}| − P

    L(π) = 0 iff every cluster lives entirely in one tier (optimal).
    L(π) > 0 means some clusters are split across tiers (bad).

Theorem (Betti-Guided Partitioning Bound):
    Let π* be the Betti-guided partition that keeps each cluster C_p
    in its entirety in the fastest available tier (HBM first, then DDR,
    then NVMe). Then for any query q:

    (i) Expected retrieval latency:
        E[T(q)] ≤ t_HBM · (N_HBM/N) + t_DDR · (N_DDR/N) + t_NVMe · (N_NVMe/N)

        where N_X = Σ_{p: π*(C_p)=X} |C_p| is the number of keys in tier X,
        and t_X is the access latency for tier X.

    (ii) Latency improvement over random partitioning:
        E[T(q)]_optimal / E[T(q)]_random ≤ 1 − (P − P_HBM)/P · (1 − t_HBM/t_NVMe)

        where P_HBM is the number of clusters fitting in HBM.

    (iii) Under the angular sparse attention model with sparsity s,
        the effective latency is further reduced:
        E[T(q)]_sparse ≤ (1 − s) · E[T(q)]_optimal

        because fraction s of blocks are pruned before any memory access.

    Proof of (i):
        1. Query q maps to a cluster C_p via Voronoi lookup (O(1) by TSS).
        2. All keys in C_p are in the same tier by construction.
        3. Access latency = t_{π*(C_p)}.
        4. Averaging over uniform query distribution:
           E[T] = Σ_p (|C_p|/N) · t_{π*(C_p)}
        5. This decomposes into the tier-weighted sum.

    Proof of (ii):
        1. Random: each key independently placed. P(key in HBM) = cap_HBM/N.
           E[T]_random = t_HBM · cap_HBM/N + t_DDR · cap_DDR/N + t_NVMe · rest/N
        2. Betti-guided: clusters sorted by access frequency (or size),
           packed greedily into HBM, then DDR, then NVMe.
        3. The improvement comes from avoiding cross-tier cluster splits.
           Each split forces at least one slow-tier access per query.
        4. With P clusters and P_HBM fitting in HBM:
           The saved slow accesses ≈ (P - P_HBM)/P fraction of queries
           that would have hit NVMe under random placement.

    Proof of (iii):
        Angular sparse attention prunes fraction s of blocks.
        These blocks are never accessed, regardless of tier.
        The effective query touches only (1-s) fraction of the cache.

    ∎

Numerical Example (H100 with DeepSeek 70B, 128K context):
    KV cache size: 128K × 2 (K+V) × 128 (head_dim) × 80 (layers) × 2 (fp16)
                 ≈ 5.24 GB per batch element
    Assume β₀ = 200 clusters, sparsity s = 0.7 (70% blocks pruned).

    Access latencies: t_HBM = 100ns, t_DDR = 300ns, t_NVMe = 50000ns.

    Random: E[T] ≈ 0.5 × 100 + 0.3 × 300 + 0.2 × 50000 = 10,140 ns
    Betti:  E[T] ≈ 0.8 × 100 + 0.2 × 300 = 140 ns (all hot clusters in HBM)
    Sparse: E[T] ≈ 0.3 × 140 = 42 ns

    Speedup: 241× over random.

═══════════════════════════════════════════════════════════════
"""
from __future__ import annotations

import math
from typing import Any, Dict, List, Optional, Tuple

import numpy as np


# Memory tier specifications
class MemoryTier:
    HBM3 = "HBM3"      # H100: 80GB, 3.35 TB/s, ~100ns latency
    DDR5 = "DDR5"       # Host: ~512GB, 204.8 GB/s, ~300ns latency
    NVME = "NVMe"       # Storage: ~8TB, 7 GB/s, ~50μs latency

TIER_LATENCY_NS = {
    MemoryTier.HBM3: 100.0,
    MemoryTier.DDR5: 300.0,
    MemoryTier.NVME: 50_000.0,
}

TIER_BANDWIDTH_GBS = {
    MemoryTier.HBM3: 3350.0,
    MemoryTier.DDR5: 204.8,
    MemoryTier.NVME: 7.0,
}

TIER_CAPACITY_GB = {
    MemoryTier.HBM3: 80.0,
    MemoryTier.DDR5: 512.0,
    MemoryTier.NVME: 8000.0,
}


def kv_cache_size_gb(seq_len: int, n_layers: int = 80,
                     head_dim: int = 128, n_kv_heads: int = 8,
                     dtype_bytes: int = 2, batch_size: int = 1) -> float:
    """
    Compute KV cache size in GB.

    Size = seq_len × 2(K+V) × head_dim × n_kv_heads × n_layers × dtype × batch
    """
    total_bytes = (seq_len * 2 * head_dim * n_kv_heads * n_layers *
                   dtype_bytes * batch_size)
    return total_bytes / (1024 ** 3)


def topological_locality(partition: List[str], cluster_ids: List[int]) -> int:
    """
    Compute topological locality L(π).

    L(π) = Σ_p |{tiers used by C_p}| − P

    Returns 0 if each cluster is entirely in one tier.
    """
    # Group by cluster
    cluster_tiers: Dict[int, set] = {}
    for tier, cid in zip(partition, cluster_ids):
        cluster_tiers.setdefault(cid, set()).add(tier)

    P = len(cluster_tiers)
    return sum(len(tiers) for tiers in cluster_tiers.values()) - P


class BettiGuidedPartitioner:
    """
    Partitions KV cache keys across memory tiers using Betti-0 clusters.

    Algorithm:
        1. Compute Betti-0 decomposition of the key space.
        2. Sort clusters by size (largest first = most queried).
        3. Greedily pack clusters into HBM, then DDR, then NVMe.
        4. Never split a cluster across tiers.
    """

    def __init__(self, hbm_capacity_gb: float = 80.0,
                 ddr_capacity_gb: float = 512.0):
        self.hbm_cap = hbm_capacity_gb
        self.ddr_cap = ddr_capacity_gb

    def partition(self, cluster_sizes: List[int],
                  bytes_per_key: int = 256) -> Dict[str, Any]:
        """
        Assign clusters to tiers.

        Parameters
        ----------
        cluster_sizes : list of int (size of each cluster)
        bytes_per_key : int (bytes per KV pair)

        Returns
        -------
        dict with tier assignments, latency analysis, locality score.
        """
        P = len(cluster_sizes)
        N = sum(cluster_sizes)

        # Sort clusters by size (descending = most important first)
        sorted_clusters = sorted(enumerate(cluster_sizes),
                                 key=lambda x: x[1], reverse=True)

        tier_assignments: Dict[int, str] = {}
        hbm_used_gb = 0.0
        ddr_used_gb = 0.0
        n_hbm, n_ddr, n_nvme = 0, 0, 0

        for cid, size in sorted_clusters:
            size_gb = size * bytes_per_key / (1024 ** 3)

            if hbm_used_gb + size_gb <= self.hbm_cap:
                tier_assignments[cid] = MemoryTier.HBM3
                hbm_used_gb += size_gb
                n_hbm += size
            elif ddr_used_gb + size_gb <= self.ddr_cap:
                tier_assignments[cid] = MemoryTier.DDR5
                ddr_used_gb += size_gb
                n_ddr += size
            else:
                tier_assignments[cid] = MemoryTier.NVME
                n_nvme += size

        # Compute expected latency (Betti-guided)
        latency_betti = (
            TIER_LATENCY_NS[MemoryTier.HBM3] * n_hbm / max(N, 1) +
            TIER_LATENCY_NS[MemoryTier.DDR5] * n_ddr / max(N, 1) +
            TIER_LATENCY_NS[MemoryTier.NVME] * n_nvme / max(N, 1)
        )

        # Compute expected latency (random partition)
        hbm_frac = self.hbm_cap / (self.hbm_cap + self.ddr_cap + 8000)
        ddr_frac = self.ddr_cap / (self.hbm_cap + self.ddr_cap + 8000)
        nvme_frac = 1.0 - hbm_frac - ddr_frac
        latency_random = (
            TIER_LATENCY_NS[MemoryTier.HBM3] * hbm_frac +
            TIER_LATENCY_NS[MemoryTier.DDR5] * ddr_frac +
            TIER_LATENCY_NS[MemoryTier.NVME] * nvme_frac
        )

        # Locality score
        partition_list = [tier_assignments.get(i, MemoryTier.NVME)
                          for i in range(P)]
        cluster_id_list = list(range(P))
        locality = topological_locality(partition_list, cluster_id_list)

        # Count P_HBM
        p_hbm = sum(1 for t in tier_assignments.values()
                     if t == MemoryTier.HBM3)

        return {
            "tier_assignments": tier_assignments,
            "n_hbm": n_hbm,
            "n_ddr": n_ddr,
            "n_nvme": n_nvme,
            "hbm_used_gb": hbm_used_gb,
            "ddr_used_gb": ddr_used_gb,
            "P_total": P,
            "P_hbm": p_hbm,
            "latency_betti_ns": latency_betti,
            "latency_random_ns": latency_random,
            "speedup": latency_random / max(latency_betti, 0.001),
            "topological_locality": locality,
            "perfect_locality": locality == 0,
        }

    def sparse_latency(self, base_latency_ns: float,
                       sparsity: float) -> float:
        """
        E[T]_sparse = (1 − s) · E[T]_base

        Angular sparse attention prunes fraction s of blocks
        before any memory access occurs.
        """
        return (1.0 - sparsity) * base_latency_ns


def h100_kv_analysis(seq_len: int = 131072, n_layers: int = 80,
                     n_clusters: int = 200, sparsity: float = 0.7) -> Dict[str, Any]:
    """
    Full H100 KV-cache analysis for DeepSeek 70B at 128K context.
    """
    cache_gb = kv_cache_size_gb(seq_len, n_layers)

    # Generate synthetic cluster sizes (Zipf distribution — realistic)
    rng = np.random.RandomState(42)
    raw = rng.zipf(1.5, n_clusters).astype(float)
    proportions = raw / raw.sum()
    N_keys = seq_len * 2  # K and V
    cluster_sizes = [max(1, int(p * N_keys)) for p in proportions]
    cluster_sizes[-1] = N_keys - sum(cluster_sizes[:-1])

    partitioner = BettiGuidedPartitioner()
    result = partitioner.partition(cluster_sizes, bytes_per_key=256)

    # Add sparsity analysis
    sparse_latency = partitioner.sparse_latency(
        result["latency_betti_ns"], sparsity
    )

    return {
        "seq_len": seq_len,
        "n_layers": n_layers,
        "kv_cache_gb": cache_gb,
        "n_clusters": n_clusters,
        "sparsity": sparsity,
        **result,
        "sparse_latency_ns": sparse_latency,
        "total_speedup": result["latency_random_ns"] / max(sparse_latency, 0.001),
    }

File: kernel/epsilon/epsilon/crates/aether-core/src/persistent_kv_partition.rs
// Epsilon-Hollow - Copyright (c) 2024 Teerth Sharma
// SPDX-License-Identifier: Epsilon-Hollow

//! Persistent Homology KV-cache Partitioning (PHKP) in `no_std` Rust.
//!
//! This replaces the legacy `epsilon_core/persistent_kv_partition.py` theorem
//! slice with fixed-capacity Rust. It keeps Betti-0 clusters intact while
//! greedily packing larger clusters into faster tiers.

/// H100 HBM3 access latency in nanoseconds.
pub const HBM3_LATENCY_NS: f64 = 100.0;
/// Host DDR5 access latency in nanoseconds.
pub const DDR5_LATENCY_NS: f64 = 300.0;
/// NVMe access latency in nanoseconds.
pub const NVME_LATENCY_NS: f64 = 50_000.0;
/// H100 HBM3 bandwidth in GB/s.
pub const HBM3_BANDWIDTH_GBS: f64 = 3350.0;
/// Host DDR5 bandwidth in GB/s.
pub const DDR5_BANDWIDTH_GBS: f64 = 204.8;
/// NVMe bandwidth in GB/s.
pub const NVME_BANDWIDTH_GBS: f64 = 7.0;
/// Default H100 HBM3 capacity in GB.
pub const HBM3_CAPACITY_GB: f64 = 80.0;
/// Default host DDR5 capacity in GB.
pub const DDR5_CAPACITY_GB: f64 = 512.0;
/// Default NVMe capacity in GB.
pub const NVME_CAPACITY_GB: f64 = 8000.0;
const BYTES_PER_GIB: f64 = 1024.0 * 1024.0 * 1024.0;

/// Memory tier used for KV-cache placement.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MemoryTier {
    /// H100 HBM3.
    Hbm3,
    /// Host DDR5.
    Ddr5,
    /// NVMe storage.
    Nvme,
}

impl MemoryTier {
    /// Tier access latency in nanoseconds.
    pub const fn latency_ns(self) -> f64 {
        match self {
            Self::Hbm3 => HBM3_LATENCY_NS,
            Self::Ddr5 => DDR5_LATENCY_NS,
            Self::Nvme => NVME_LATENCY_NS,
        }
    }

    /// Tier bandwidth in GB/s.
    pub const fn bandwidth_gbs(self) -> f64 {
        match self {
            Self::Hbm3 => HBM3_BANDWIDTH_GBS,
            Self::Ddr5 => DDR5_BANDWIDTH_GBS,
            Self::Nvme => NVME_BANDWIDTH_GBS,
        }
    }
}

/// Compute KV-cache size in GiB.
pub fn kv_cache_size_gb(
    seq_len: u64,
    n_layers: u32,
    head_dim: u32,
    n_kv_heads: u32,
    dtype_bytes: u32,
    batch_size: u32,
) -> f64 {
    let total_bytes = seq_len as f64
        * 2.0
        * f64::from(head_dim)
        * f64::from(n_kv_heads)
        * f64::from(n_layers)
        * f64::from(dtype_bytes)
        * f64::from(batch_size);
    total_bytes / BYTES_PER_GIB
}

/// Compute topological locality `L(pi)` for a partition and cluster-id stream.
///
/// `0` means each cluster lives entirely in one tier. A positive value means
/// one or more Betti-0 clusters were split across tiers.
pub fn topological_locality(partition: &[MemoryTier], cluster_ids: &[usize]) -> usize {
    let len = partition.len().min(cluster_ids.len());
    let mut locality = 0_usize;

    for idx in 0..len {
        let cid = cluster_ids[idx];
        let mut first_seen = true;
        if cluster_ids[..idx].iter().any(|&c| c == cid) {
            first_seen = false;
        }
        if !first_seen {
            continue;
        }

        let mut tiers = [false; 3];
        for scan in 0..len {
            if cluster_ids[scan] == cid {
                tiers[tier_index(partition[scan])] = true;
            }
        }
        let tier_count = tiers.iter().filter(|used| **used).count();
        locality += tier_count.saturating_sub(1);
    }

    locality
}

/// Greedy Betti-guided KV-cache partitioner.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct BettiGuidedPartitioner {
    hbm_capacity_gb: f64,
    ddr_capacity_gb: f64,
}

impl Default for BettiGuidedPartitioner {
    fn default() -> Self {
        Self::new(HBM3_CAPACITY_GB, DDR5_CAPACITY_GB)
    }
}

impl BettiGuidedPartitioner {
    /// Create a partitioner with explicit HBM and DDR capacities in GiB.
    pub const fn new(hbm_capacity_gb: f64, ddr_capacity_gb: f64) -> Self {
        Self {
            hbm_capacity_gb,
            ddr_capacity_gb,
        }
    }

    /// HBM capacity in GiB.
    pub const fn hbm_capacity_gb(&self) -> f64 {
        self.hbm_capacity_gb
    }

    /// DDR capacity in GiB.
    pub const fn ddr_capacity_gb(&self) -> f64 {
        self.ddr_capacity_gb
    }

    /// Assign each Betti-0 cluster to one memory tier without splitting it.
    pub fn partition<const N: usize>(
        &self,
        cluster_sizes: &[u64; N],
        bytes_per_key: u64,
    ) -> PartitionReport<N> {
        let mut order = [0_usize; N];
        for (idx, slot) in order.iter_mut().enumerate() {
            *slot = idx;
        }
        sort_indices_by_size_desc(&mut order, cluster_sizes);

        let mut assignments = [MemoryTier::Nvme; N];
        let mut hbm_used_gb = 0.0;
        let mut ddr_used_gb = 0.0;
        let mut n_hbm = 0_u64;
        let mut n_ddr = 0_u64;
        let mut n_nvme = 0_u64;

        for &cluster_idx in &order {
            let size = cluster_sizes[cluster_idx];
            let size_gb = size as f64 * bytes_per_key as f64 / BYTES_PER_GIB;
            if hbm_used_gb + size_gb <= self.hbm_capacity_gb {
                assignments[cluster_idx] = MemoryTier::Hbm3;
                hbm_used_gb += size_gb;
                n_hbm += size;
            } else if ddr_used_gb + size_gb <= self.ddr_capacity_gb {
                assignments[cluster_idx] = MemoryTier::Ddr5;
                ddr_used_gb += size_gb;
                n_ddr += size;
            } else {
                assignments[cluster_idx] = MemoryTier::Nvme;
                n_nvme += size;
            }
        }

        let total_keys = cluster_sizes.iter().copied().sum::<u64>();
        let denom = total_keys.max(1) as f64;
        let latency_betti_ns = HBM3_LATENCY_NS * n_hbm as f64 / denom
            + DDR5_LATENCY_NS * n_ddr as f64 / denom
            + NVME_LATENCY_NS * n_nvme as f64 / denom;

        let total_capacity = self.hbm_capacity_gb + self.ddr_capacity_gb + NVME_CAPACITY_GB;
        let hbm_frac = self.hbm_capacity_gb / total_capacity;
        let ddr_frac = self.ddr_capacity_gb / total_capacity;
        let nvme_frac = 1.0 - hbm_frac - ddr_frac;
        let latency_random_ns =
            HBM3_LATENCY_NS * hbm_frac + DDR5_LATENCY_NS * ddr_frac + NVME_LATENCY_NS * nvme_frac;

        let mut cluster_ids = [0_usize; N];
        for (idx, slot) in cluster_ids.iter_mut().enumerate() {
            *slot = idx;
        }
        let locality = topological_locality(&assignments, &cluster_ids);
        let p_hbm = assignments
            .iter()
            .filter(|tier| matches!(tier, MemoryTier::Hbm3))
            .count();

        PartitionReport {
            assignments,
            n_hbm,
            n_ddr,
            n_nvme,
            hbm_used_gb,
            ddr_used_gb,
            p_total: N,
            p_hbm,
            latency_betti_ns,
            latency_random_ns,
            speedup: latency_random_ns / latency_betti_ns.max(0.001),
            topological_locality: locality,
            perfect_locality: locality == 0,
        }
    }

    /// Sparse-attention latency bound after pruning fraction `sparsity`.
    pub fn sparse_latency(&self, base_latency_ns: f64, sparsity: f64) -> f64 {
        (1.0 - sparsity.clamp(0.0, 1.0)) * base_latency_ns
    }
}

/// PHKP partition report for fixed-size cluster arrays.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct PartitionReport<const N: usize> {
    /// Tier assignment for each cluster.
    pub assignments: [MemoryTier; N],
    /// Keys placed in HBM3.
    pub n_hbm: u64,
    /// Keys placed in DDR5.
    pub n_ddr: u64,
    /// Keys placed in NVMe.
    pub n_nvme: u64,
    /// HBM3 capacity consumed in GiB.
    pub hbm_used_gb: f64,
    /// DDR5 capacity consumed in GiB.
    pub ddr_used_gb: f64,
    /// Total cluster count.
    pub p_total: usize,
    /// Number of clusters placed in HBM3.
    pub p_hbm: usize,
    /// Expected Betti-guided latency in nanoseconds.
    pub latency_betti_ns: f64,
    /// Expected random-placement latency in nanoseconds.
    pub latency_random_ns: f64,
    /// Random latency divided by Betti-guided latency.
    pub speedup: f64,
    /// Locality penalty. Zero means no cluster is split across tiers.
    pub topological_locality: usize,
    /// True when `topological_locality == 0`.
    pub perfect_locality: bool,
}

fn tier_index(tier: MemoryTier) -> usize {
    match tier {
        MemoryTier::Hbm3 => 0,
        MemoryTier::Ddr5 => 1,
        MemoryTier::Nvme => 2,
    }
}

fn sort_indices_by_size_desc<const N: usize>(order: &mut [usize; N], sizes: &[u64; N]) {
    for i in 0..N {
        let mut best = i;
        for j in (i + 1)..N {
            if sizes[order[j]] > sizes[order[best]] {
                best = j;
            }
        }
        if best != i {
            order.swap(i, best);
        }
    }
}
